File: backend/models/space_container_standard.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Dict, Mapping

from .space_container import SpaceContainer


def _utc_now() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def _normalize_spec_status(status: str) -> str:
    value = str(status or "").strip().upper()
    if value == "PASS":
        return "completed"
    if value == "FAIL":
        return "failed"
    if value == "RUNNING":
        return "running"
    return "pending"


@dataclass
class StandardGeoReference:
    station: str
    coord_system: str
    coords: Dict[str, float]
    gps: Dict[str, float]
    alignment: str

    def to_dict(self) -> Dict[str, Any]:
        return {
            "station": self.station,
            "coord_system": self.coord_system,
            "coords": dict(self.coords),
            "gps": dict(self.gps),
            "alignment": self.alignment,
        }


@dataclass
class StandardApplicableSpec:
    spuId: str
    status: str
    attempts: int
    latest_node: str | None
    depends_on: list[str] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "spuId": self.spuId,
            "status": self.status,
            "attempts": self.attempts,
            "latest_node": self.latest_node,
            "depends_on": list(self.depends_on),
        }


@dataclass
class StandardNormExecution:
    applicable_specs: list[StandardApplicableSpec] = field(default_factory=list)
    current_state: str = "draft"
    gate_status: str = "awaiting_lab"
    execution_order: list[str] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "applicable_specs": [item.to_dict() for item in self.applicable_specs],
            "current_state": self.current_state,
            "gate_status": self.gate_status,
            "execution_order": list(self.execution_order),
        }


@dataclass
class StandardRuntime:
    active_spec: str | None = None
    active_form: str = ""
    pending_actions: list[str] = field(default_factory=list)
    pending_signatures: list[str] = field(default_factory=list)
    last_action: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "active_spec": self.active_spec,
            "active_form": self.active_form,
            "pending_actions": list(self.pending_actions),
            "pending_signatures": list(self.pending_signatures),
            "last_action": self.last_action,
        }


@dataclass
class StandardLifecycle:
    state: str = "ACTIVE"
    created_at: str = ""
    updated_at: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "state": self.state,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
        }


@dataclass
class SpaceContainerStandard:
    v_address: str
    container_type: str
    geo_reference: StandardGeoReference
    norm_execution: StandardNormExecution
    runtime: StandardRuntime
    lifecycle: StandardLifecycle

    def to_dict(self) -> Dict[str, Any]:
        return {
            "v_address": self.v_address,
            "container_type": self.container_type,
            "geo_reference": self.geo_reference.to_dict(),
            "norm_execution": self.norm_execution.to_dict(),
            "runtime": self.runtime.to_dict(),
            "lifecycle": self.lifecycle.to_dict(),
        }


def build_standard_space_container(
    container: SpaceContainer,
    *,
    slot_payload: Mapping[str, Any] | None = None,
    node_history: list[Mapping[str, Any]] | None = None,
) -> SpaceContainerStandard:
    slot_geo: Mapping[str, Any] = {}
    if isinstance(slot_payload, Mapping):
        maybe_geo = slot_payload.get("geo")
        if isinstance(maybe_geo, Mapping):
            slot_geo = maybe_geo

    station = str(slot_geo.get("station", "")).strip() or "K19+070"
    x = float(slot_geo.get("x", 0.0) or 0.0)
    y = float(slot_geo.get("y", 0.0) or 0.0)
    z = float(slot_geo.get("elevation", 0.0) or 0.0)
    alignment = str(slot_geo.get("alignment", "")).strip()

    execution_order = list(container.norm_execution.specs_bound)
    applicable_specs: list[StandardApplicableSpec] = []
    for index, spu_id in enumerate(execution_order):
        binding = next((item for item in container.spec_bindings if item.spuId == spu_id), None)
        attempts = 0
        if binding is not None:
            attempts = sum(
                1
                for item in (node_history or [])
                if str(item.get("spu_id", "")).strip() == spu_id
            )
        depends_on = [execution_order[index - 1]] if index > 0 else []
        applicable_specs.append(
            StandardApplicableSpec(
                spuId=spu_id,
                status=_normalize_spec_status(binding.status if binding is not None else "DRAFT"),
                attempts=attempts,
                latest_node=binding.latest_node if binding is not None else None,
                depends_on=depends_on,
            )
        )

    active_binding = next((item for item in container.spec_bindings if item.status == "RUNNING"), None)
    if active_binding is None:
        active_binding = next((item for item in container.spec_bindings if item.status == "DRAFT"), None)

    gate_status = "awaiting_lab"
    if any(item.status == "FAIL" for item in container.spec_bindings):
        gate_status = "needs_retest"
    elif any(item.status == "RUNNING" for item in container.spec_bindings):
        gate_status = "in_execution"
    elif container.lifecycle_state == "ARCHIVED":
        gate_status = "archived"

    history_times: list[str] = []
    for item in (node_history or []):
        created_at = str(item.get("created_at", "")).strip()
        updated_at = str(item.get("completed_at") or item.get("created_at") or "").strip()
        if created_at:
            history_times.append(created_at)
        if updated_at:
            history_times.append(updated_at)
    history_times.sort()
    created_at = history_times[0] if history_times else _utc_now()
    updated_at = history_times[-1] if history_times else _utc_now()

    return SpaceContainerStandard(
        v_address=container.v_address,
        container_type="space",
        geo_reference=StandardGeoReference(
            station=station,
            coord_system="CGCS2000",
            coords={"X": x, "Y": y, "Z": z},
            gps={"lat": 0.0, "lng": 0.0},
            alignment=alignment,
        ),
        norm_execution=StandardNormExecution(
            applicable_specs=applicable_specs,
            current_state=str(container.lifecycle_state or "DRAFT").lower(),
            gate_status=gate_status,
            execution_order=execution_order,
        ),
        runtime=StandardRuntime(
            active_spec=active_binding.spuId if active_binding is not None else None,
            active_form=active_binding.spuId if active_binding is not None else "",
            pending_actions=["fill_form", "submit_test"] if container.lifecycle_state != "ARCHIVED" else [],
            pending_signatures=[],
            last_action=updated_at,
        ),
        lifecycle=StandardLifecycle(
            state="ARCHIVED" if container.lifecycle_state == "ARCHIVED" else "ACTIVE",
            created_at=created_at,
            updated_at=updated_at,
        ),
    )

File: apps/executable-spec-web/src/platform/spatial/space-container-standard.ts
import type {
  ExecutionNode,
  SchedulerTaskStatus,
  SpaceContainer,
  SpaceContainerApplicableSpec,
  SpaceContainerGeoReference,
  SpaceContainerStandardModel,
  SpaceSlot,
} from "../types.ts";

type RuntimeSnapshot = {
  currentSpuId: string | null;
  currentNodeId: string | null;
  phase: "idle" | "running" | "signing" | "completed";
};

function nowIso(): string {
  return new Date().toISOString();
}

function runtimeSnapshot(container: SpaceContainer): RuntimeSnapshot {
  const runtime = (container as SpaceContainer & { runtime?: Partial<RuntimeSnapshot> }).runtime;
  return {
    currentSpuId: typeof runtime?.currentSpuId === "string" ? runtime.currentSpuId : null,
    currentNodeId: typeof runtime?.currentNodeId === "string" ? runtime.currentNodeId : null,
    phase: runtime?.phase ?? "idle",
  };
}

function normalizeTaskStatus(status: "DRAFT" | "RUNNING" | "PASS" | "FAIL"): SchedulerTaskStatus {
  if (status === "PASS") {
    return "completed";
  }
  if (status === "FAIL") {
    return "failed";
  }
  if (status === "RUNNING") {
    return "running";
  }
  return "pending";
}

function computeExecutionOrder(container: SpaceContainer): string[] {
  return container.specBindings.map((item) => item.spuId);
}

function computeDependencies(executionOrder: string[]): Map<string, string[]> {
  const map = new Map<string, string[]>();
  executionOrder.forEach((spuId, index) => {
    if (index === 0) {
      map.set(spuId, []);
      return;
    }
    map.set(spuId, [executionOrder[index - 1]]);
  });
  return map;
}

function geoReferenceFromSlot(slot: SpaceSlot | null): SpaceContainerGeoReference {
  if (!slot) {
    return {
      station: "K19+070",
      coordSystem: "CGCS2000",
      coords: { X: 0, Y: 0, Z: 0 },
      gps: { lat: 0, lng: 0 },
      alignment: "",
    };
  }
  return {
    station: slot.geo.station,
    coordSystem: "CGCS2000",
    coords: {
      X: Number(slot.geo.x),
      Y: Number(slot.geo.y),
      Z: Number(slot.geo.elevation),
    },
    gps: {
      lat: 0,
      lng: 0,
    },
    alignment: slot.geo.alignment ?? "",
  };
}

function resolveActiveNode(container: SpaceContainer, nodes: ExecutionNode[], runtime: RuntimeSnapshot): ExecutionNode | null {
  if (runtime.currentNodeId) {
    const currentNode = nodes.find((node) => node.nodeId === runtime.currentNodeId) ?? null;
    if (currentNode) {
      return currentNode;
    }
  }
  const activeBinding = runtime.currentSpuId
    ? container.specBindings.find((item) => item.spuId === runtime.currentSpuId) ?? null
    : container.specBindings.find((item) => item.status === "RUNNING")
      ?? container.specBindings.find((item) => item.status === "DRAFT")
      ?? null;
  if (!activeBinding?.latestNodeId) {
    return null;
  }
  return nodes.find((node) => node.nodeId === activeBinding.latestNodeId) ?? null;
}

function inferGateStatus(container: SpaceContainer): string {
  if (container.lifecycleState === "ARCHIVED") {
    return "archived";
  }
  if (container.specBindings.some((binding) => binding.status === "FAIL")) {
    return "needs_retest";
  }
  if (container.specBindings.some((binding) => binding.status === "RUNNING")) {
    return "in_execution";
  }
  return "awaiting_lab";
}

function inferPendingActions(container: SpaceContainer, activeNode: ExecutionNode | null): string[] {
  if (container.lifecycleState === "ARCHIVED") {
    return [];
  }
  if (!activeNode) {
    return ["create_node", "fill_form", "submit_test"];
  }
  if (activeNode.status === "SIGNING") {
    return ["sign", "finalize_node"];
  }
  if (activeNode.status === "DRAFT") {
    return ["fill_form", "submit_test"];
  }
  if (activeNode.status === "RUNNING") {
    return ["wait_execution"];
  }
  return ["create_node"];
}

function latestActionTime(nodes: ExecutionNode[]): string {
  const candidates = nodes
    .map((item) => item.updatedAt)
    .filter((item) => typeof item === "string" && item.trim().length > 0)
    .sort();
  return candidates[candidates.length - 1] ?? nowIso();
}

function createdTime(nodes: ExecutionNode[]): string {
  const candidates = nodes
    .map((item) => item.createdAt)
    .filter((item) => typeof item === "string" && item.trim().length > 0)
    .sort();
  return candidates[0] ?? nowIso();
}

export function toStandardSpaceContainer(
  container: SpaceContainer,
  options?: {
    slot?: SpaceSlot | null;
    nodes?: ExecutionNode[];
  },
): SpaceContainerStandardModel {
  const slot = options?.slot ?? null;
  const nodes = options?.nodes ?? [];
  const runtime = runtimeSnapshot(container);
  const executionOrder = computeExecutionOrder(container);
  const dependsOnMap = computeDependencies(executionOrder);

  const applicableSpecs: SpaceContainerApplicableSpec[] = container.specBindings.map((binding) => ({
    spuId: binding.spuId,
    status: normalizeTaskStatus(binding.status),
    attempts: binding.historyNodeIds.length,
    latestNode: binding.latestNodeId ?? null,
    dependsOn: dependsOnMap.get(binding.spuId) ?? [],
  }));

  const activeSpecBinding = runtime.currentSpuId
    ? container.specBindings.find((item) => item.spuId === runtime.currentSpuId) ?? null
    : container.specBindings.find((item) => item.status === "RUNNING")
      ?? container.specBindings.find((item) => item.status === "DRAFT")
      ?? null;
  const activeNode = resolveActiveNode(container, nodes, runtime);

  return {
    vAddress: container.vAddress,
    containerType: "space",
    geoReference: geoReferenceFromSlot(slot),
    normExecution: {
      applicableSpecs,
      currentState: container.lifecycleState.toLowerCase(),
      gateStatus: inferGateStatus(container),
      executionOrder,
    },
    runtime: {
      activeSpec: activeSpecBinding?.spuId ?? null,
      activeForm: activeSpecBinding?.spuId ? `FORM_${activeSpecBinding.spuId.replace(/[^A-Za-z0-9]+/g, "_").toUpperCase()}` : "",
      pendingActions: inferPendingActions(container, activeNode),
      pendingSignatures: activeNode
        ? activeNode.requiredSignatures.filter((role) => !activeNode.signedBy.includes(role))
        : [],
      lastAction: latestActionTime(nodes),
    },
    lifecycle: {
      state: container.lifecycleState === "ARCHIVED" ? "ARCHIVED" : "ACTIVE",
      createdAt: createdTime(nodes),
      updatedAt: latestActionTime(nodes),
    },
  };
}
